### PR TITLE
#285 milk predictions

### LINKED ISSUE #285: Provide predictions for expected milk yield
Hi, 

Could TestDayResultEventResource have a new property:   "milkWeightExpected" or "milkWeightGoal" ? 

BR

### PR DESCRIPTION
**Pull request by erwinspeybroeck**
Changed accordingly the remarks

### CODE REVIEW

File: resources/icarMilkPredictionResource.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+          "$ref": "../types/icarMilkingPredictionType.json"
+        },
+        "daysInMilkAtLactationPeak": {
+          "type": "integer",
+          "description": "The days in milk in a lactation when the peak production is expected to occur."
+        },
+        "lactationPeakProduction": {
+          "$ref": "../types/icarMilkingMilkWeightType.json"
+        },
+        "PredictedProductionNextMR": {
```
Comment (cookeac) [inline resources/icarMilkPredictionResource.json]:
```suggestion
        "productionNextMilkRecording": {
```

File: resources/icarMilkPredictionResource.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  "description": "This returns a prediction of the expected milk, fat and protein for a cow for the remaining the lactation. Also a prediction of this production on the next testday is included and it eventually gives indication of when the peak production is reached.",
+
+  "allOf": [{
+      "$ref": "../resources/icarAnimalEventCoreResource.json"
+    },
+    {
+      "type": "object",
+
+      "properties": {
+        "averagePredictedProduction": {
```
Comment (cookeac) [inline resources/icarMilkPredictionResource.json]:
```suggestion
        "averageRemainingProduction": {
```

File: resources/icarTestDayResultEventResource.json [hunk 3]
```diff
@@ -19,7 +19,10 @@
           "items": {
             "$ref": "../types/icarMilkCharacteristicsType.json"
           }
-        }
+        },
+        "PredictedProductionOnTestday": {
```
Comment (cookeac) [inline resources/icarTestDayResultEventResource.json]:
```suggestion
        "predictedProductionOnTestday": {
```

File: types/icarMilkingFatWeightType.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (cookeac) [inline types/icarMilkingFatWeightType.json]: Remove this file.

File: types/icarMilkingPredictionType.json [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+  "type": "object",
+
+  "required": [
+    "unitCode",
+    "value"
+  ],
+
+  "properties": {
+    "MilkWeight": {
```
Comment (cookeac) [inline types/icarMilkingPredictionType.json]: Consider using icarMassMeasureType
Also check capitalisation.

**Review by cookeac — CHANGES_REQUESTED** [review]
@erwinspeybroeck to make the changes.

File: types/icarMilkingPredictionType.json [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      "$ref": "../types/icarMilkingMilkWeightType.json"
```
Comment (erwinspeybroeck) [inline types/icarMilkingPredictionType.json]:
```suggestion
      "$ref": "../types/icarMassMeasureType.json"
```

File: types/icarMilkingPredictionType.json [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    },
+    "FatWeight": {
+      "$ref": "../types/icarMilkingFatWeightType.json"
```
Comment (erwinspeybroeck) [inline types/icarMilkingPredictionType.json]:
```suggestion
      "$ref": "../types/icarMassMeasureType.json"
```

File: types/icarMilkingPredictionType.json [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    },
+    "ProteinWeight": {
+      "$ref": "../types/icarMilkingProteinWeightType.json"
```
Comment (erwinspeybroeck) [inline types/icarMilkingPredictionType.json]:
```suggestion
      "$ref": "../types/icarMassMeasureType.json"
```

File: resources/icarMilkPredictionResource.json [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+      "properties": {
+        "averageRemainingProduction": {
+          "$ref": "../types/icarMilkingPredictionType.json"
+        },
+        "daysInMilkAtLactationPeak": {
+          "type": "integer",
+          "description": "The days in milk in a lactation when the peak production is expected to occur."
+        },
+        "lactationPeakProduction": {
+          "$ref": "../types/icarMilkingMilkWeightType.json"
```
Comment (erwinspeybroeck) [inline resources/icarMilkPredictionResource.json]:
```suggestion
          "$ref": "../types/icarMassMeasureType.json"
```

File: types/icarMilkingPredictionType.json (hunk 5 above)
Comment (cookeac) [inline types/icarMilkingPredictionType.json]:
```suggestion
    "milkWeight": {
```

File: types/icarMilkingPredictionType.json [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+  "required": [
+    "unitCode",
+    "value"
+  ],
+
+  "properties": {
+    "MilkWeight": {
+      "$ref": "../types/icarMassMeasureType.json"
+    },
+    "FatWeight": {
```
Comment (cookeac) [inline types/icarMilkingPredictionType.json]:
```suggestion
    "fatWeight": {
```

File: types/icarMilkingPredictionType.json [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+      "$ref": "../types/icarMassMeasureType.json"
+    },
+    "ProteinWeight": {
```
Comment (cookeac) [inline types/icarMilkingPredictionType.json]:
```suggestion
    "proteinWeight": {
```

File: types/icarMilkingPredictionType.json [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+      "$ref": "../types/icarMassMeasureType.json"
+    },
+    "Hours": {
```
Comment (cookeac) [inline types/icarMilkingPredictionType.json]:
```suggestion
    "hours": {
```

File: url-schemes/exampleUrlScheme.json [hunk 13]
```diff
@@ -14,7 +14,7 @@
     "url": "https://icar-ade.standard.com"
   }],
   "tags": [{
-      "name": "ADE-1.3 additions",
+      "name": "ADE-1.4 additions",
```
Comment (cookeac) [inline url-schemes/exampleUrlScheme.json]:
```suggestion
      "name": "ADE-1.3 additions",
```

**Review by cookeac — CHANGES_REQUESTED** [review]
@erwinspeybroeck In addition to the few suggestions I have added in the files (you should be able to just accept/commit these), please consider:
1. Removing the changes to `icarStatisticsResource` and `icarStatisticsGroupType` as these are now corrected in the published ADE 1.3.2. Depending on how you are editing, you might choose to "stash" the files with the milk prediction changes, then refresh your branch from the Develop branch, and then put your milk prediction files back in.
2. In the URL schemes, for the moment just leave all the tags as "-1.3-" rather than 1.3.2 or 1.4. We will update all the tags in bulk before we publish version 1.4.

Thanks

File: resources/icarStatisticsResource.json [hunk 14]
```diff
@@ -34,13 +34,7 @@
...
       "type": "array",
       "items": {
-        "$ref": "../types/icarStatisticsGroupType.json"
-      }
-    },
-    "statistics": {
-      "type": "array",
-      "items": {
-        "$ref": "../types/icarStatisticsType.json"
+        "$ref": "../types/icarStatisticsGroupType.json",
```
Comment (erwinspeybroeck) [inline resources/icarStatisticsResource.json]:
```suggestion
        "$ref": "../types/icarStatisticsGroupType.json"
```

File: url-schemes/exampleUrlScheme.json [hunk 15]
```diff
@@ -29,6 +29,11 @@
       "name": "ADE-1.3-management",
       "description": "Management messages approved by the working group"
     },
+    {
+      "name": "ADE-1.3.2-management",
```
Comment (erwinspeybroeck) [inline url-schemes/exampleUrlScheme.json]:
```suggestion
      "name": "ADE-1.3-management",
```

File: url-schemes/exampleUrlScheme.json [hunk 16]
```diff
@@ -249,6 +254,45 @@
         }
       }
     },
+    "/locations/{location-scheme}/{location-id}/milk-predictions": {
+        "get": {
+            "operationId": "get-milk-predictions",
+            "summary": "Get the milk predictions for a certain location",
+            "description": "# Purpose\nProvides the milk predictions for a location\n",
+            "tags": [
+                "ADE-1.4 additions"
```
Comment (erwinspeybroeck) [inline url-schemes/exampleUrlScheme.json]:
```suggestion
                "ADE-1.3 additions"
```

File: url-schemes/exampleUrlScheme.json [hunk 17]
```diff
@@ -680,7 +724,7 @@
                 "summary": "Get the statistics for a certain location",
                 "description": "# Purpose\nProvides the statistics for a location\n One can specify the period for which the statistics can be calculated by specifying a date to and date from parameter. Also the purpose can be provided to limit the purpose of the statistics",
                 "tags": [
-                    "ADE-1.3-management"
+                    "ADE-1.3.2-management"
```
Comment (erwinspeybroeck) [inline url-schemes/exampleUrlScheme.json]:
```suggestion
                    "ADE-1.3-management"
```

File: url-schemes/managementURLScheme.json [hunk 18]
```diff
@@ -19,6 +19,10 @@
         {
             "name": "ADE-1.3-management",
             "description": "Management messages approved by the working group"
+        },
+        {
+            "name": "ADE-1.3.2-management",
```
Comment (erwinspeybroeck) [inline url-schemes/managementURLScheme.json]:
```suggestion
            "name": "ADE-1.3-management",
```

File: url-schemes/managementURLScheme.json [hunk 19]
```diff
@@ -448,7 +452,7 @@
                 "summary": "Get the statistics for a certain location",
                 "description": "# Purpose\nProvides the statistics for a location\n One can specify the period for which the statistics can be calculated by specifying a date to and date from parameter. Also the purpose can be provided to limit the purpose of the statistics",
                 "tags": [
-                    "ADE-1.3-management"
+                    "ADE-1.3.2-management"
```
Comment (erwinspeybroeck) [inline url-schemes/managementURLScheme.json]:
```suggestion
                    "ADE-1.3-management"
```

File: url-schemes/milkURLScheme.json [hunk 20]
```diff
@@ -16,6 +16,10 @@
         }
     ],
     "tags": [
+        {
+          "name": "ADE-1.4-milk",
```
Comment (erwinspeybroeck) [inline url-schemes/milkURLScheme.json:20]:
```suggestion
          "name": "ADE-1.3-milk",
```

File: url-schemes/milkURLScheme.json [hunk 21]
```diff
@@ -376,6 +380,45 @@
                 }
             }
         },
+        "/locations/{location-scheme}/{location-id}/milk-predictions": {
+            "get": {
+                "operationId": "get-milk-predictions",
+                "summary": "Get the milk predictions for a certain location",
+                "description": "# Purpose\nProvides the milk predictions for a location\n",
+                "tags": [
+                    "ADE-1.4-milk"
```
Comment (erwinspeybroeck) [inline url-schemes/milkURLScheme.json]:
```suggestion
                    "ADE-1.3-milk"
```

File: url-schemes/milkURLScheme.json [hunk 22]
```diff
@@ -16,6 +16,10 @@
         }
     ],
     "tags": [
+        {
+          "name": "ADE-1.3-milk",
```
Comment (erwinspeybroeck) [inline url-schemes/milkURLScheme.json]:
```suggestion
          "name": "ADE-1.4-milk",
```

File: url-schemes/managementURLScheme.json [hunk 23]
```diff
@@ -16,6 +16,10 @@
         }
     ],
     "tags": [
+        {
+            "name": "ADE-1.3-management",
```
Comment (erwinspeybroeck) [inline url-schemes/managementURLScheme.json]:
```suggestion
            "name": "ADE-1.4-management",
```

**Comment by erwinspeybroeck** [comment]
will not use it

**Pull request closed without merging**